File: java/src/com/algos/topics/li13_graph/Dfs_L332_ReconstructItinerary.java
package  com.algos.topics.li13_graph;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class Dfs_L332_ReconstructItinerary {

	public static void main(String[] args) {
		Dfs_L332_ReconstructItinerary obj = new Dfs_L332_ReconstructItinerary();

		String[][] tickets = {
			{ "MUC", "LHR" },
			{ "JFK", "MUC" },
			{ "SFO", "SJC" },
			{ "LHR", "SFO" } };
		List<List<String>> list = new ArrayList<>();
		for(String[] ticket: tickets) {
			list.add(new ArrayList<String>(Arrays.asList(ticket)));
		}
		System.out.println(obj.findItinerary(list));
	}

	// Problem: https://leetcode.com/problems/reconstruct-itinerary/
	// Idea: We need to find path in tree/graph where path contains all edges in order.
	// Since, we need to give preference to arivals in lex order, we need to sort the tickets in that order and build graph adj list in that order.
	// We need to consider each edge only once. Hence edge will be removed from graph once it is considered.
	// If that path doesn't give solution, we backtrack and add that edge back.
	// We can't use vis arrau for nodes as same node can be visited multiple times.
	public List<String> findItinerary(List<List<String>> tickets) {
		Collections.sort(tickets, (a, b)->{
			if(a.get(0).equals(b.get(0))) {
				return a.get(1).compareTo(b.get(1));
			}else {
				return a.get(0).compareTo(b.get(0));
			}
		});

		Map<String, List<String>> adj = new HashMap<>();
		for(List<String> ticket: tickets) {
			String src = ticket.get(0);
			adj.putIfAbsent(src, new ArrayList<>());
			adj.get(src).add(ticket.get(1));
		}

		List<String> ans = new ArrayList<>();
		ans.add("JFK");
		dfs(adj, "JFK", ans, tickets.size());
		return ans;
	}

	private boolean dfs(Map<String, List<String>> adj, String src, List<String> ans, int length) {
		if(ans.size() == length+1)
			return true;

		if(!adj.containsKey(src)) 				// if no outgoing edge from a node and we still didn't reach n+1 nodes above
			return false;

		for(int i=0; i<adj.get(src).size(); i++) {
			String dst = adj.get(src).get(i);
			ans.add(dst);
			adj.get(src).remove(i);
			if(dfs(adj, dst, ans, length))
				return true;

			adj.get(src).add(i, dst);
			ans.remove(ans.size()-1);
		}

		return false;
	}

}

File: dsa/concepts/13_graphs/python/dfs_l332_reconstruct_itinerary.py
"""
Problem:
https://leetcode.com/problems/reconstruct-itinerary/

Idea:
- We must use all tickets exactly once (edge used once)
- Always choose lexicographically smallest destination
- Sort tickets first and build adjacency list in sorted order
- Use DFS with backtracking
- Remove an edge once used; if path fails, restore it

Time: O(E^2) worst case due to backtracking
Space: O(E)
"""


def find_itinerary(tickets):
    # Sort tickets lexicographically by source, then destination
    tickets.sort(key=lambda x: (x[0], x[1]))

    adj = {}
    for src, dst in tickets:
        adj.setdefault(src, []).append(dst)

    path = ["JFK"]
    dfs(adj, "JFK", path, len(tickets))
    return path


"""
DFS to build itinerary using all tickets exactly once
"""


def dfs(adj, src, path, total_edges):
    if len(path) == total_edges + 1:
        return True

    if src not in adj:
        return False

    for i in range(len(adj[src])):
        dst = adj[src][i]

        path.append(dst)
        adj[src].pop(i)

        if dfs(adj, dst, path, total_edges):
            return True

        # backtrack
        adj[src].insert(i, dst)
        path.pop()

    return False


# -------------------- TEST --------------------
if __name__ == "__main__":
    tickets = [
        ["MUC", "LHR"],
        ["JFK", "MUC"],
        ["SFO", "SJC"],
        ["LHR", "SFO"]
    ]

    print(find_itinerary(tickets))
